torch/csrc/jit/passes/quantization.cpp

```diff
@@ -716,6 +719,150 @@ c10::optional<IValue> toTwoElementIntList(Value* v) {
+          if (module_set_.insert(m._ivalue()).second) {
+            uses_to_rewrite_.push_back(instance);
+            GRAPH_DEBUG("Found use to rewrite: ", instance);
+          }
+        } else {
+          GRAPH_DEBUG(
+              "Can't handle the access pattern of GetAttr ",
+              "in make submodule uses unique, traced back to ",
+              boundary_val->debugName(),
+              " which is not self: ",
+              self->debugName());
+        }
+      }
+    }
+  }
+
+  // Deduplicate module uses given the information we recorded before
+  void dedupModuleUses() {
+    for (Value* v : uses_to_rewrite_) {
+      const auto& path = value_to_path_map_.at(v);
+      const auto& m = findChildModule(module_, path);
+      // add a clone of the child module to the parent of the duplicated module
+      const auto& child_name = addChildModule(module_, m, path);
+      TORCH_INTERNAL_ASSERT(v->node()->kind() == prim::GetAttr);
+      // change the name in GetAttr call
+      auto original_name = v->node()->s(attr::name);
+      v->node()->s_(attr::name, child_name);
+      GRAPH_UPDATE(
+          "Module use dedup: changing use of original module ",
+          original_name,
+          " to ",
+          child_name);
+    }
+  }
+
+  script::Module findChildModule(
+      const script::Module& module,
+      const std::vector<std::string>& path) {
+    script::Module m = module;
+    for (size_t i = 0; i < path.size(); ++i) {
+      m = m.attr(path[i]).toModule();
+    }
+    return m;
+  }
+
+  std::string addChildModule(
+      script::Module module,
+      const script::Module& child_module,
+      const std::vector<std::string>& path) {
+    TORCH_INTERNAL_ASSERT(
+        path.size() > 0, "path must have at least one element.");
+    // Parent module of the leaf child module corresponding to
+    // the path
+    auto parent_of_leaf = findChildModule(
+        module, std::vector<std::string>(path.begin(), path.end() - 1));
+
+    // Original name of the child module
+    std::string original_name = path[path.size() - 1];
+    int uid = 0;
+    std::string child_name = original_name + "_" + c10::to_string(uid++);
+    while (parent_of_leaf.hasattr(child_name)) {
+      child_name = original_name + "_" + c10::to_string(uid++);
+    }
+    parent_of_leaf.register_module(child_name, child_module.clone_instance());
+    return child_name;
+  }
+
+  // Check if value is the input of the graph
+  bool hitGraphInput(Value* value) {
+    Graph* graph = value->owningGraph();
+    const auto& inputs = graph->inputs();
+    return std::find(inputs.begin(), inputs.end(), value) != inputs.end();
+  }
+
+  Value* getModuleAccessPath(Value* instance, std::vector<std::string>& path) {
+    // Iterator to traverse back the GetAttr calls
+    Value* iter = instance;
+    // trace back the instance to recover the path of the submodule
+    while (!hitGraphInput(iter) && iter->node()->kind() == prim::GetAttr) {
```

Comment:
Is the assumption here that if you hit a graph input, you've hit the `self` value? This is going to break for modules that are passed in as arguments, for example:

```
import torch

class Nest1(torch.nn.Module):
    def forward(self, x, my_activation):
        return my_activation(x) + my_activation(x)

class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.relu = torch.nn.ReLU()
        self.nest1 = Nest1()

    def forward(self, x):
        return self.nest1(x, self.relu)


f = Foo()
traced = torch.jit.trace(f, (torch.rand(3, 4),))

traced._c.dump(True, False, False)

torch._C._jit_pass_dedup_module_uses(traced._c)

traced._c.dump(True, False, False)
```

(and yes, people do do this, especially for activation modules)

I'll think through a design for how we can do this correctly in the general case

Reply:
No, this case is taken care of, after we hit graph input we'll check if the value is equal to self, see L761

Reply:
we don't support this use in this PR, is there a need to support this?

Reply:
It's not a good idea to check in this pass if it creates an ill-formed Module, as in my example above

Reply:
I'm not sure I understand, what do you mean?

Reply:
For your example there should be no effect to the module. 
I just found a bug that reverses the check of whether we should copy the use or not, it is fixed now.